File: samples/akka-sample-cluster-java/src/main/java/sample/cluster/transformation/Frontend.java
/*
 * Copyright (C) 2009-2024 Lightbend Inc. <https://www.lightbend.com>
 */
package sample.cluster.transformation;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.*;
import akka.actor.typed.receptionist.Receptionist;

//#frontend
public class Frontend extends AbstractBehavior<Frontend.Event> {

  interface Event {}
  private enum Tick implements Event {
    INSTANCE
  }
  private static final class WorkersUpdated implements Event {
    public final Set<ActorRef<Worker.TransformText>> newWorkers;
    public WorkersUpdated(Set<ActorRef<Worker.TransformText>> newWorkers) {
      this.newWorkers = newWorkers;
    }
  }
  private static final class TransformCompleted implements Event {
    public final String originalText;
    public final String transformedText;
    public TransformCompleted(String originalText, String transformedText) {
      this.originalText = originalText;
      this.transformedText = transformedText;
    }
  }
  private static final class JobFailed implements Event {
    public final String why;
    public final String text;
    public JobFailed(String why, String text) {
      this.why = why;
      this.text = text;
    }
  }

  public static Behavior<Event> create() {
    return Behaviors.setup(context ->
        Behaviors.withTimers(timers ->
          new Frontend(context, timers)
        )
    );
  }

  private final List<ActorRef<Worker.TransformText>> workers = new ArrayList<>();
  private int jobCounter = 0;

  private Frontend(ActorContext<Event> context, TimerScheduler<Event> timers) {
    super(context);
    ActorRef<Receptionist.Listing> subscriptionAdapter =
        context.messageAdapter(Receptionist.Listing.class, listing ->
          new WorkersUpdated(listing.getServiceInstances(Worker.WORKER_SERVICE_KEY)));
    context.getSystem().receptionist().tell(Receptionist.subscribe(Worker.WORKER_SERVICE_KEY, subscriptionAdapter));

    timers.startTimerWithFixedDelay(Tick.INSTANCE, Tick.INSTANCE, Duration.ofSeconds(2));
  }

  @Override
  public Receive<Event> createReceive() {
    return newReceiveBuilder()
        .onMessage(WorkersUpdated.class, this::onWorkersUpdated)
        .onMessage(TransformCompleted.class, this::onTransformCompleted)
        .onMessage(JobFailed.class, this::onJobFailed)
        .onMessageEquals(Tick.INSTANCE, this::onTick)
        .build();
  }


  private Behavior<Event> onTransformCompleted(TransformCompleted event) {
    getContext().getLog().info("Got completed transform of {}: {}", event.originalText, event.transformedText);
    return this;
  }

  private Behavior<Event> onJobFailed(JobFailed event) {
    getContext().getLog().warn("Transformation of text {} failed. Because: {}", event.text, event.why);
    return this;
  }

  private Behavior<Event> onTick() {
    if (workers.isEmpty()) {
      getContext().getLog().warn("Got tick request but no workers available, not sending any work");
    } else {
      // how much time can pass before we consider a request failed
      Duration timeout = Duration.ofSeconds(5);
      ActorRef<Worker.TransformText> selectedWorker = workers.get(jobCounter % workers.size());
      getContext().getLog().info("Sending work for processing to {}", selectedWorker);
      String text = "hello-" + jobCounter;
      getContext().ask(
          Worker.TextTransformed.class,
          selectedWorker,
          timeout,
          responseRef -> new Worker.TransformText(text, responseRef),
          (response, failure) -> {
            if (response != null) {
              return new TransformCompleted(text, response.text);
            } else {
              return new JobFailed("Processing timed out", text);
            }
          }
      );
      jobCounter++;
    }
    return this;
  }

  private Behavior<Event> onWorkersUpdated(WorkersUpdated event) {
    workers.clear();
    workers.addAll(event.newWorkers);
    getContext().getLog().info("List of services registered with the receptionist changed: {}", event.newWorkers);
    return this;
  }
}
//#frontend

File: samples/akka-sample-cluster-scala/src/main/scala/sample/cluster/transformation/Frontend.scala
/*
 * Copyright (C) 2009-2025 Lightbend Inc. <https://www.lightbend.com>
 */
package sample.cluster.transformation

import scala.concurrent.duration._
import akka.util.Timeout
import akka.actor.typed.ActorRef
import akka.actor.typed.Behavior
import akka.actor.typed.receptionist.Receptionist
import akka.actor.typed.scaladsl.ActorContext
import akka.actor.typed.scaladsl.Behaviors
import sample.cluster.transformation.Worker.TextTransformed

import scala.util.Failure
import scala.util.Success

//#frontend
object Frontend {

  sealed trait Event
  private case object Tick extends Event
  private final case class WorkersUpdated(newWorkers: Set[ActorRef[Worker.TransformText]]) extends Event
  private final case class TransformCompleted(originalText: String, transformedText: String) extends Event
  private final case class JobFailed(why: String, text: String) extends Event

  def apply(): Behavior[Event] = Behaviors.setup { ctx =>
    Behaviors.withTimers { timers =>
      // subscribe to available workers
      val subscriptionAdapter = ctx.messageAdapter[Receptionist.Listing] {
        case Worker.WorkerServiceKey.Listing(workers) =>
          WorkersUpdated(workers)
      }
      ctx.system.receptionist ! Receptionist.Subscribe(Worker.WorkerServiceKey, subscriptionAdapter)

      timers.startTimerWithFixedDelay(Tick, Tick, 2.seconds)

      running(ctx, IndexedSeq.empty, jobCounter = 0)
    }
  }

  private def running(
      ctx: ActorContext[Event],
      workers: IndexedSeq[ActorRef[Worker.TransformText]],
      jobCounter: Int): Behavior[Event] =
    Behaviors.receiveMessage {
      case WorkersUpdated(newWorkers) =>
        ctx.log.info("List of services registered with the receptionist changed: {}", newWorkers)
        running(ctx, newWorkers.toIndexedSeq, jobCounter)
      case Tick =>
        if (workers.isEmpty) {
          ctx.log.warn("Got tick request but no workers available, not sending any work")
          Behaviors.same
        } else {
          // how much time can pass before we consider a request failed
          implicit val timeout: Timeout = 5.seconds
          val selectedWorker = workers(jobCounter % workers.size)
          ctx.log.info("Sending work for processing to {}", selectedWorker)
          val text = s"hello-$jobCounter"
          // FIXME why does Scala 3 need the explicit type on replyTo to not infer reply to AnyRef
          ctx.ask(selectedWorker, (replyTo: ActorRef[TextTransformed]) => Worker.TransformText(text, replyTo)) {
            case Success(transformedText) =>
              TransformCompleted(originalText = text, transformedText.text)
            case Failure(_) => JobFailed("Processing timed out", text)
          }
          running(ctx, workers, jobCounter + 1)
        }
      case TransformCompleted(originalText, transformedText) =>
        ctx.log.info("Got completed transform of {}: {}", originalText, transformedText)
        Behaviors.same

      case JobFailed(why, text) =>
        ctx.log.warn("Transformation of text {} failed. Because: {}", text, why)
        Behaviors.same

    }
}
//#frontend
